New Tutorial - moveit_py motion planning python api tutorial (#565)
ros-planning/moveit2_tutorials: doc/examples/motion_planning_python_api/motion_planning_python_api_tutorial.rst

Motion Planning Python API
==================================

\textbf{Video tutorial to be added soon}
.. raw:: html

        <iframe width="560" height="315" src="https://www.youtube.com/embed/7KvF7Dj7bz0" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

In this tutorial we are going to cover the basics of the motion planning API for moveit_py. This tutorial is broken down into the following sections:

* **Getting Started:** An outline of the tutorial setup requirements.
* **Understanding Planning Parameters:** An outline of setting parameters for supported planners.
* **Single Pipeline Planning (Default Configurations):** Planning using prespecified robot configurations. 
* **Single Pipeline Planning (Robot State):** Planning using a robot state instance.
* **Single Pipeline Planning (Pose Goal):** Planning using a pose goal.
* **Single Pipeline Planning (Custom Constraints):** Planning using custom constraints.
* **Multi Pipeline Planning:** Running multiple planning pipelines in parallel.

> The code for this tutorial can be found `here <https://github.com/peterdavidfagan/moveit2_tutorials/tree/moveit_py_motion_planning_python_api_tutorial/doc/examples/motion_planning_python_api>`_. 

Author: [6bfc4de](https://github.com/ros-planning/moveit2_tutorials/pull/565/commits/6bfc4deb8d899200ddc90970eda5e5e12ae67b07)

> The moveit_py API is documented `here <>`_.

Author: The second link isn't populated as we don't have hosted API docs yet for the Python library. 

The following pull request is open for this: https://github.com/ros-planning/moveit2_tutorials/pull/504

> The moveit_py API is documented `here <>`_.

Author: [6bfc4de](https://github.com/ros-planning/moveit2_tutorials/pull/565/commits/6bfc4deb8d899200ddc90970eda5e5e12ae67b07)

Getting Started
==================================

> In order to complete this tutorial, you must have set up a workspace that includes MoveIt2 and its corresponding tutorials. An excellent outline on how to set up such a workspace is provided in the :doc:`Getting Started Guide </doc/tutorials/getting_started/getting_started>`, please consult this guide for further information.

Author: [6bfc4de](https://github.com/ros-planning/moveit2_tutorials/pull/565/commits/6bfc4deb8d899200ddc90970eda5e5e12ae67b07)

> Once you have setup your workspace, you can execute the code for this tutorial by running the following command: ::

Author: [e1667b0](https://github.com/ros-planning/moveit2_tutorials/pull/564/commits/e1667b027a7ff466839b2a27da7e9d5daa7a8ce4)

> Once you have setup your workspace, you can execute the code for this tutorial by running the following command: ::

Author: [6bfc4de](https://github.com/ros-planning/moveit2_tutorials/pull/565/commits/6bfc4deb8d899200ddc90970eda5e5e12ae67b07)

>         ros2 launch moveit2_tutorials motion_planning_python_api.launch.py

Author: [00ff463](https://github.com/ros-planning/moveit2_tutorials/pull/565/commits/00ff463f222ed358de627ecff907f625d0974db6)

Understanding Planning Parameters
==================================

> MoveIt supports the use of multiple planning libraries out of the box, it is important that we provide settings/parameters for the planners we wish to use. To accomplish this with moveit_py we specify a yaml configuration file that defines the parameters associated with our moveit_py node. An example of such a configuration file is given below: ::

Author: [00ff463](https://github.com/ros-planning/moveit2_tutorials/pull/565/commits/00ff463f222ed358de627ecff907f625d0974db6)

                        planning_time: 0.5

        ompl_rrt_star:
                plan_request_params:
                        planning_attempts: 1
                        planning_pipeline: ompl_rrt_star # Different OMPL pipeline name!
                        planner_id: "RRTstarkConfigDefault"
                        max_velocity_scaling_factor: 1.0
                        max_acceleration_scaling_factor: 1.0
                        planning_time: 1.5

        pilz_lin:
                plan_request_params:
                        planning_attempts: 1
                        planning_pipeline: pilz_industrial_motion_planner
                        planner_id: "PTP"
                        max_velocity_scaling_factor: 1.0
                        max_acceleration_scaling_factor: 1.0
                        planning_time: 0.8

        chomp:
                plan_request_params:
                        planning_attempts: 1
                        planning_pipeline: chomp
                        max_velocity_scaling_factor: 1.0
                        max_acceleration_scaling_factor: 1.0
                        planning_time: 1.5

> The first block of the configuration file sets the planning scene monitor options such as the topics that it subsribes to (Note: if you aren't familiar with the planning scene monitor, you should consider reviewing ... ): ::

Reviewer: Missing a link to the planning scene monitor doc you intended

        planning_scene_monitor_options:
                name: "planning_scene_monitor"
                robot_description: "robot_description"
                joint_state_topic: "/joint_states"
                attached_collision_object_topic: "/moveit_cpp/planning_scene_monitor"
                publish_planning_scene_topic: "/moveit_cpp/publish_planning_scene"
                monitored_planning_scene_topic: "/moveit_cpp/monitored_planning_scene"
                wait_for_initial_state_timeout: 10.0

> The second block of the configuration file sets the planning pipelines that we wish to use. MoveIt supports multiple motion planning libraries including OMPL, PILZ industrial motion planner, Stochastic Trajectory Optimization for Motion Planning (STOMP), Search-Based Planning Library (SBPL) and Covariant Hamiltonian Optimization for Motion Planning (CHOMP) to name a few. When configuring our ``moveit_py`` node, we need to specify the configuration for planning pipelines we wish to use: ::

Reviewer: Is there a link to a doc section listing all the planners we can throw here?

Maybe https://moveit.picknik.ai/humble/doc/examples/examples.html#configuration

                        planning_attempts: 1
                        planning_pipeline: ompl_rrt_star
                        planner_id: "RRTstarkConfigDefault"
                        max_velocity_scaling_factor: 1.0
                        max_acceleration_scaling_factor: 1.0
                        planning_time: 1.5

        pilz_lin:
                plan_request_params:
                        planning_attempts: 1
                        planning_pipeline: pilz_industrial_motion_planner
                        planner_id: "PTP"
                        max_velocity_scaling_factor: 1.0
                        max_acceleration_scaling_factor: 1.0
                        planning_time: 0.8

        chomp:
                plan_request_params:
                        planning_attempts: 1
                        planning_pipeline: chomp
                        max_velocity_scaling_factor: 1.0
                        max_acceleration_scaling_factor: 1.0
                        planning_time: 1.5

These specified parameters will be made available as ``moveit_py`` node parameters and will be leveraged at runtime when performing planning. This is what we will investigate next.

> Instatiating moveit_py and planning component

Reviewer: ```suggestion
Instantiating moveit_py and planning component
```

        # instantiate a RobotState instance using the current robot model
        robot_model = moveit.get_robot_model()
        robot_state = RobotState(robot_model)

        # randomize the robot state
        robot_state.set_to_random_positions()

        # set plan start state to current state
        panda_arm.set_start_state_to_current_state()

        # set goal state to the initialized robot state
        logger.info("Set goal state to the initialized robot state")
        panda_arm.set_goal(robot_state=robot_state)

        # plan to goal
        logger.info("Planning trajectory")
        plan_result = panda_arm.plan()

        # execute the plan
        if plan_result:
                logger.info("Executing plan")
                panda_arm.execute()

Single Pipeline Planning - Pose Goal
==================================================

> Another common way to specify a goal state is via a pose goal ros message. Here we demonstrate how to set a pose goal for the end effector of the robot: ::

Reviewer: ```suggestion
Another common way to specify a goal state is via a pose goal ROS message. Here we demonstrate how to set a pose goal for the end effector of the robot: ::
```

        # set plan start state to current state
        panda_arm.set_start_state_to_current_state()

        # set pose goal with PoseStamped message
        pose_goal = PoseStamped()
        pose_goal.header.frame_id = "panda_link0"
        pose_goal.pose.orientation.w = 1.0
        pose_goal.pose.position.x = 0.28
        pose_goal.pose.position.y = -0.2
        pose_goal.pose.position.z = 0.5
        pose_goal = {"link_name": "panda_link8", "pose": pose_goal}
        panda_arm.set_goal(pose_goal=pose_goal)

        # plan to goal
        logger.info("Planning trajectory")
        plan_result = panda_arm.plan()

        # execute the plan
        if plan_result:
                logger.info("Executing plan")
                panda_arm.execute()

Single Pipeline Planning - Custom Constraints
==================================================

> Arguably the most flexible way to specify a goal which we wish to plan for is via custom constraints, here we demonstrate planning to a configuration that satisfies a set of joint constraints: ::

Reviewer: I'd leave out the subjective wording.

```suggestion
You can also control the output of motion planning via custom constraints. Here we demonstrate planning to a configuration that satisfies a set of joint constraints: ::
```

If there is a doc link to constraints, would be nice to have here too.

        # set plan start state to current state
        panda_arm.set_start_state_to_current_state()

        # set plan start state to current state
        panda_arm.set_start_state_to_current_state()

        # set constraints message
        joint_values = {
                "panda_joint1": -1.0,
                "panda_joint2": 0.7,
                "panda_joint3": 0.7,
                "panda_joint4": -1.5,
                "panda_joint5": -0.7,
                "panda_joint6": 2.0,
                "panda_joint7": 0.0,
        }
        robot_state.joint_positions = joint_values
        joint_constraint = construct_joint_constraint(
                robot_state=robot_state,
                joint_model_group=moveit.get_robot_model().get_joint_model_group("panda_arm"),
        )
        panda_arm.set_goal(motion_plan_constraints=[joint_constraint])

        # plan to goal
        logger.info("Planning trajectory")
        plan_result = panda_arm.plan()

        # execute the plan
        if plan_result:
                logger.info("Executing plan")
                panda_arm.execute()

Multi Pipeline Planning
===========================

> A recent addition to ``moveit_cpp`` and ``moveit_py`` is the ability to simultaneously execute multiple planning pipelines and select the resulting motion plan amongst all generated motion plans that best satisfies your task requirements. In previous sections, we defined a set of planning pipelines, here we will see how we can leverage these settings to simultaneously performing planning with multiple pipelines : ::

Reviewer: I think using "parallel" instead of "simultaneous" is more precise and true to the new parallel planning feature. Something like this?

```suggestion
A recent addition to ``moveit_cpp`` and ``moveit_py`` is the ability to execute multiple planning pipelines in parallel and select the resulting motion plan amongst all generated motion plans that best satisfies your task requirements.
In previous sections, we defined a set of planning pipelines.
Here we will see how to plan in parallel with several of these pipelines : ::
```
